src/Features/Core/Portable/Completion/Providers/AbstractTypeImportCompletionProvider.cs

```diff
@@ -223,14 +232,32 @@ internal override async Task<CompletionChange> GetChangeAsync(Document document,
 
                 return CompletionChange.Create(Utilities.Collapse(newText, builder.ToImmutableAndFree()));
             }
-            else
+
+            static bool ShouldCompleteWithFullyQualifyTypeName(Document document)
             {
-                // For workspace that doesn't support document change, e.g. DebuggerIntellisense
-                // we complete the type name in its fully qualified form instead.
-                var fullyQualifiedName = $"{containingNamespace}.{completionItem.DisplayText}";
-                var change = new TextChange(completionListSpan, fullyQualifiedName);
+                var workspace = document.Project.Solution.Workspace;
 
-                return CompletionChange.Create(change);
+                // Certain types of workspace don't support document change, e.g. DebuggerIntellisense
+                if (!workspace.CanApplyChange(ApplyChangesKind.ChangeDocument))
+                {
+                    return true;
+                }
+
+                // During an EnC session, adding import is not supported.
+                var encService = workspace.Services.GetService<IDebuggingWorkspaceService>()?.EditAndContinueServiceOpt;
+                if (encService?.EditSession != null)
+                {
+                    return true;
+                }
+
+                // Certain documents, e.g. Razor document, don't support adding imports
+                var documentSupportsFeatureService = workspace.Services.GetService<IDocumentSupportsFeatureService>();
```

Comment:
hmmm.. why not add one specific for type import?

Reply:
I think the existing `SupportRefactoring` already does what I need, so I just used it. Do you prefer adding one specific to adding import? That feels too narrow to me.